cl/opinion_page/types.py

```diff
@@ -0,0 +1,80 @@
+from dataclasses import dataclass
+from typing import List
+
+from django.urls import reverse
+
+from cl.search.models import OpinionCluster, OpinionsCitedByRECAPDocument
+
+
+@dataclass
+class ViewAuthority:
+    caption: str
+    count: int
+    url: str
+
+    @classmethod
+    def from_cluster_authority(cls, auth: OpinionCluster, query_string: str):
+        opinion_url = f"{auth.get_absolute_url()}?{query_string}"
+        citation_count = auth.citation_depth
+        opinion_caption = auth.caption
+        return cls(
+            caption=opinion_caption, count=citation_count, url=opinion_url
+        )
+
+    @classmethod
+    def from_recap_cit_record(
+        cls, record: OpinionsCitedByRECAPDocument, query_string: str
+    ):
+        opinion_caption = record.cited_opinion.cluster.caption
+        citation_count = record.depth
+        opinion_url = (
+            f"{record.cited_opinion.cluster.get_absolute_url()}?{query_string}"
+        )
+        return cls(
+            caption=opinion_caption, count=citation_count, url=opinion_url
+        )
+
+
+@dataclass
+class AuthoritiesContext:
+    total_authorities_count: int
+    top_authorities: List[ViewAuthority]
+    view_all_url: str
+
+    @classmethod
+    def from_opinion_cluster(
+        cls, cluster: OpinionCluster, request_query_string: str
+    ):
+        top_authorities = [
+            ViewAuthority.from_cluster_authority(ca, request_query_string)
+            for ca in cluster.authorities_with_data[:5]
+        ]
+        view_all_url_base = reverse(
+            "view_authorities", args=[cluster.pk, cluster.slug]
+        )
+        return cls(
+            top_authorities=top_authorities,
+            view_all_url=f"{view_all_url_base}?{request_query_string}",
+            total_authorities_count=len(cluster.authorities_with_data),
+        )
+
+    @classmethod
+    def from_recap_document_cits(
+        cls,
+        total_cit_count: int,
+        cit_records: List[OpinionsCitedByRECAPDocument],
+        request_query_string: str,
+    ):
+        top_authorities = [
+            ViewAuthority.from_recap_cit_record(
+                r, query_string=request_query_string
+            )
+            for r in cit_records
+        ]
```

Comment:
I _think_ this means that we make a bunch more queries than before, but I could be wrong. It could be that the queries were just better hidden in the template and they're more obvious in the view. I do see that in your screenshot, the page took 20 queries to load, which feels like a lot. I suspect this accounts for a bunch of them.

In any case, now that we see them, I wonder if we could refactor this to use a `select_related` or `prefetch_related` for this and the same pattern you have above?

Reply:
It's only 2 queries in the function that gets the citation records for a recap doc: https://github.com/freelawproject/courtlistener/pull/3317/files#diff-a8a8b6a59162fd73eea97d6b21c17656610182629a4302eaaa54d5432b1a9c62R63

Confirmed by expanding the queries debug thingy.

Some other stuff in the view function seems to be generating extra queries that are susceptible to optimization, e.g., https://github.com/freelawproject/courtlistener/blob/main/cl/opinion_page/views.py#L342 (`select_releated`) 

There are some lookups of the citation table by `cluster_id` too, but it's not immediately clear to me where those are coming from tbh

Reply:
> I think this means that we make a bunch more queries than before, but I could be wrong.

@mlissner You're right! Excluding the authorities section reduces the number of queries required to load the page to just **5**. Here's a screenshot of the Django debug toolbar when we load the page using the code from the `main` branch:

![image](https://github.com/freelawproject/courtlistener/assets/55959657/ac588bff-e049-4b3f-8db3-be0deec8dda7)

Using the `select_related` and `prefetch_related` methods and carefully selecting only the essential fields from the query, I successfully reduced the number of database queries to **7**. Here's a screenshot of the debug toolbar after the refactor:

<img width="1146" alt="Screenshot 2023-11-30 at 1 48 17 PM" src="https://github.com/freelawproject/courtlistener/assets/55959657/1019b853-d449-4ce7-87e2-152d23bd34f2">